### PR TITLE
Elastic byte buffer for S3OutputStream (base on master)

### PR DESCRIPTION
**Pull request by hsp8712**
## Problem

Same PR as in: https://github.com/confluentinc/kafka-connect-storage-cloud/pull/419
But rebase on master and do some updates.

##### Does this solution apply anywhere else?
- [ ] yes
- [* ] no

##### Testing done:
- [ *] Unit tests
- [ ] Integration tests
- [ ] System tests
- [ ] Manual tests

### CODE REVIEW

File: kafka-connect-s3/src/main/java/io/confluent/connect/s3/storage/SimpleByteBuf.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+ */
+public class SimpleByteBuf implements ByteBuf {
+
+  private ByteBuffer buffer;
+
+  public SimpleByteBuf(int capacity) {
+    this.buffer = ByteBuffer.allocate(capacity);
+  }
+
+  @Override
```
Comment (sudeshwasnik) [inline kafka-connect-s3/src/main/java/io/confluent/connect/s3/storage/SimpleByteBuf.java:31]: nit - @Override isn't required, we are not extending, but implementing an interface.

**Review by sudeshwasnik — COMMENTED** [review]
NICE!!! thank you for the amazing effort! 

I just noticed some very small nits, adding those now. 

NIT - `SimpleByteBuf` -> `SimpleByteBuffer`. (consistent with ElasticByteBuffer)

File: kafka-connect-s3/src/test/java/io/confluent/connect/s3/storage/ElasticByteBufferTest.java [hunk 2]
```diff
@@ -0,0 +1,327 @@
+package io.confluent.connect.s3.storage;
+
+import org.apache.commons.lang.RandomStringUtils;
+import org.apache.commons.lang.math.RandomUtils;
+import org.junit.Test;
+import java.nio.BufferOverflowException;
+import static org.junit.Assert.*;
```
Comment (sudeshwasnik) [inline kafka-connect-s3/src/test/java/io/confluent/connect/s3/storage/ElasticByteBufferTest.java:7]: super small nit- .* import 
```suggestion
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotSame;
import static org.junit.Assert.assertSame;
import static org.junit.Assert.assertThrows;
import static org.junit.Assert.assertTrue;
```

File: kafka-connect-s3/src/main/java/io/confluent/connect/s3/storage/SimpleByteBuf.java (hunk 1 above)
Comment (hsp8712) [reply kafka-connect-s3/src/main/java/io/confluent/connect/s3/storage/SimpleByteBuf.java]: Done

**Comment by sudeshwasnik** [comment]
lgtm! 
build is failing. I'll check it out and approve if it looks unrelated. thanks!

**Review by sudeshwasnik — APPROVED** [review]
Thanks for the PR! 
LGTM! 

1. Lets close all the previous PRs 
2. Will release a `MINOR` version of the connector. Will update documentation with newer configuration (and link it here too).

**Comment by hsp8712** [comment]
Great! All the previous PRs have been closed. Looking forward to release.

**Comment by sudeshwasnik** [comment]
Cool. Merging !

**Comment by sudeshwasnik** [comment]
S3Sink 10.1.0 version was released 🎉 
Here are the docs - https://docs.confluent.io/kafka-connect-s3-sink/current/configuration_options.html 
and the configuration doc added are -

```
s3.elastic.buffer.enable
Specifies whether or not to allocate elastic buffer for staging s3-part to save memory. Note that this may cause decreased performance or increased CPU usage.

Type: boolean
Default: false
Importance: low


s3.elastic.buffer.init.capacity
Elastic buffer initial capacity. Only applicable when s3.part.size is set true.

Type: int
Default: 131072 (128 KB)
Valid Values: [4096,…]
Importance: low
```

**Comment by sudeshwasnik** [comment]
@hsp8712 a new version of S3Sink is released. 
Here are the doc changes -  https://docs.confluent.io/kafka-connect-s3-sink/current/configuration_options.html 
```
s3.elastic.buffer.enable
Specifies whether or not to allocate elastic buffer for staging s3-part to save memory. Note that this may cause decreased performance or increased CPU usage.

Type: boolean
Default: false
Importance: low


s3.elastic.buffer.init.capacity
Elastic buffer initial capacity. Only applicable when s3.elastic.buffer.enable is set true.

Type: int
Default: 131072 (128 KB)
Valid Values: [4096,…]
Importance: low
```

**Comment by hsp8712** [comment]
Thanks, that's great news!

**Comment by raphaelauv** [comment]
congratulation @hsp8712 , thanks for your work !

**Pull request merged**